File: src/core/KartShared/MonoBehaviour.lua
-- Roblox MonoBehaviour等效框架
-- 用于模拟Unity的MonoBehaviour生命周期
local MonoBehaviour = {}
MonoBehaviour.__index = MonoBehaviour

-- 创建新的MonoBehaviour实例
function MonoBehaviour.new(scriptName)
    local self = setmetatable({}, MonoBehaviour)
    self.scriptName = scriptName or "UnknownScript"
    self.gameObject = nil  -- 对应的Part或Model
    self.transform = nil   -- Transform组件
    self.enabled = true
    self.started = false
    self.destroyed = false
    
    -- 生命周期状态
    self.awakeCallbacks = {}
    self.startCallbacks = {}
    self.updateCallbacks = {}
    self.fixedUpdateCallbacks = {}
    self.lateUpdateCallbacks = {}
    self.destroyCallbacks = {}
    
    -- Roblox连接
    self._heartbeatConn = nil
    self._steppedConn = nil
    
    -- 自动绑定到Roblox事件循环
    self:BindToRoblox()
    
    -- 延迟执行Awake，确保对象完全初始化
    task.defer(function()
        self:Awake()
    end)
    
    return self
end

-- Awake - 在对象创建时立即调用
function MonoBehaviour:Awake()
    for _, callback in pairs(self.awakeCallbacks) do
        callback(self)
    end
end

-- Start - 在第一帧Update前调用
function MonoBehaviour:Start()
    if not self.started then
        self.started = true
        for _, callback in pairs(self.startCallbacks) do
            callback(self)
        end
    end
end

-- Update - 每帧调用
function MonoBehaviour:Update(deltaTime)
    if not self.enabled or self.destroyed then return end
    if not self.started then self:Start() end
    
    for _, callback in pairs(self.updateCallbacks) do
        callback(self, deltaTime)
    end
end

-- FixedUpdate - 固定时间间隔调用
function MonoBehaviour:FixedUpdate(fixedDeltaTime)
    if not self.enabled or self.destroyed then return end
    
    for _, callback in pairs(self.fixedUpdateCallbacks) do
        callback(self, fixedDeltaTime)
    end
end

-- LateUpdate - 在Update后调用
function MonoBehaviour:LateUpdate(deltaTime)
    if not self.enabled or self.destroyed then return end
    
    for _, callback in pairs(self.lateUpdateCallbacks) do
        callback(self, deltaTime)
    end
end

-- OnDestroy - 对象销毁时调用
function MonoBehaviour:OnDestroy()
    self.destroyed = true
    
    -- 断开Roblox连接
    if self._heartbeatConn then
        self._heartbeatConn:Disconnect()
        self._heartbeatConn = nil
    end
    if self._steppedConn then
        self._steppedConn:Disconnect()
        self._steppedConn = nil
    end
    
    for _, callback in pairs(self.destroyCallbacks) do
        callback(self)
    end
end

-- 绑定到 Roblox 事件循环
function MonoBehaviour:BindToRoblox()
    local RunService = game:GetService("RunService")
    
    -- 确保之前的连接已经断开
    if self._heartbeatConn then
        self._heartbeatConn:Disconnect()
    end
    if self._steppedConn then
        self._steppedConn:Disconnect()
    end
    
    self._heartbeatConn = RunService.Heartbeat:Connect(function(dt)
        if self.enabled and not self.destroyed then
            self:Update(dt)
            self:LateUpdate(dt)
        end
    end)
    
    self._steppedConn = RunService.Stepped:Connect(function(_, fixedDt)
        if self.enabled and not self.destroyed then
            self:FixedUpdate(fixedDt)
        end
    end)
end

-- 启用/禁用MonoBehaviour
function MonoBehaviour:SetEnabled(enabled)
    self.enabled = enabled
end

-- 销毁MonoBehaviour
function MonoBehaviour:Destroy()
    if not self.destroyed then
        self:OnDestroy()
    end
end

-- 添加生命周期回调
function MonoBehaviour:AddAwakeCallback(callback)
    table.insert(self.awakeCallbacks, callback)
end

function MonoBehaviour:AddStartCallback(callback)
    table.insert(self.startCallbacks, callback)
end

function MonoBehaviour:AddUpdateCallback(callback)
    table.insert(self.updateCallbacks, callback)
end

function MonoBehaviour:AddFixedUpdateCallback(callback)
    table.insert(self.fixedUpdateCallbacks, callback)
end

function MonoBehaviour:AddLateUpdateCallback(callback)
    table.insert(self.lateUpdateCallbacks, callback)
end

function MonoBehaviour:AddDestroyCallback(callback)
    table.insert(self.destroyCallbacks, callback)
end

-- 设置GameObject
function MonoBehaviour:SetGameObject(robloxObject)
    local RobloxUnityAdapter = require(script.Parent.RobloxUnityAdapter)
    self.gameObject = RobloxUnityAdapter.GameObject.new(robloxObject)
    self.transform = self.gameObject.transform
end

-- 获取组件（模拟Unity的GetComponent）
function MonoBehaviour:GetComponent(componentType)
    if self.gameObject then
        return self.gameObject:GetComponent(componentType)
    end
    return nil
end

-- 获取子对象中的组件
function MonoBehaviour:GetComponentInChildren(componentType)
    if self.gameObject then
        return self.gameObject:GetComponentInChildren(componentType)
    end
    return nil
end

-- 获取子对象中的所有组件
function MonoBehaviour:GetComponentsInChildren(componentType)
    if self.gameObject then
        return self.gameObject:GetComponentsInChildren(componentType)
    end
    return {}
end

-- 添加组件
function MonoBehaviour:AddComponent(componentType)
    if self.gameObject then
        return self.gameObject:AddComponent(componentType)
    end
    return nil
end

return MonoBehaviour

File: src/shared/kartEngine/KartShared/MonoBehaviour.ts
import { RunService } from "@rbxts/services";
import RobloxUnityAdapter, { GameObject, Transform } from "./RobloxUnityAdapter";

export class MonoBehaviour {
    public scriptName: string;
    public gameObject: GameObject | undefined;
    public transform: Transform | undefined;
    public enabled: boolean;
    public started: boolean;
    public destroyed: boolean;
    
    public awakeCallbacks: ((self: MonoBehaviour) => void)[] = [];
    public startCallbacks: ((self: MonoBehaviour) => void)[] = [];
    public updateCallbacks: ((self: MonoBehaviour, deltaTime: number) => void)[] = [];
    public fixedUpdateCallbacks: ((self: MonoBehaviour, fixedDeltaTime: number) => void)[] = [];
    public lateUpdateCallbacks: ((self: MonoBehaviour, deltaTime: number) => void)[] = [];
    public destroyCallbacks: ((self: MonoBehaviour) => void)[] = [];
    
    public _heartbeatConn: RBXScriptConnection | undefined;
    public _steppedConn: RBXScriptConnection | undefined;
    
    constructor(scriptName?: string) {
        this.scriptName = scriptName || "UnknownScript";
        this.gameObject = undefined;
        this.transform = undefined;
        this.enabled = true;
        this.started = false;
        this.destroyed = false;
        
        this.BindToRoblox();
        
        task.defer(() => {
            this.Awake();
        });
    }
    
    public Awake(): void {
        for (const callback of this.awakeCallbacks) {
            callback(this);
        }
    }
    
    public Start(): void {
        if (!this.started) {
            this.started = true;
            for (const callback of this.startCallbacks) {
                callback(this);
            }
        }
    }
    
    public Update(deltaTime: number): void {
        if (!this.enabled || this.destroyed) return;
        if (!this.started) this.Start();
        
        for (const callback of this.updateCallbacks) {
            callback(this, deltaTime);
        }
    }
    
    public FixedUpdate(fixedDeltaTime: number): void {
        if (!this.enabled || this.destroyed) return;
        
        for (const callback of this.fixedUpdateCallbacks) {
            callback(this, fixedDeltaTime);
        }
    }
    
    public LateUpdate(deltaTime: number): void {
        if (!this.enabled || this.destroyed) return;
        
        for (const callback of this.lateUpdateCallbacks) {
            callback(this, deltaTime);
        }
    }
    
    public OnDestroy(): void {
        this.destroyed = true;
        
        if (this._heartbeatConn) {
            this._heartbeatConn.Disconnect();
            this._heartbeatConn = undefined;
        }
        if (this._steppedConn) {
            this._steppedConn.Disconnect();
            this._steppedConn = undefined;
        }
        
        for (const callback of this.destroyCallbacks) {
            callback(this);
        }
    }
    
    public BindToRoblox(): void {
        if (this._heartbeatConn) {
            this._heartbeatConn.Disconnect();
        }
        if (this._steppedConn) {
            this._steppedConn.Disconnect();
        }
        
        this._heartbeatConn = RunService.Heartbeat.Connect((dt) => {
            if (this.enabled && !this.destroyed) {
                this.Update(dt);
                this.LateUpdate(dt);
            }
        });
        
        this._steppedConn = RunService.Stepped.Connect((_, fixedDt) => {
            if (this.enabled && !this.destroyed) {
                this.FixedUpdate(fixedDt);
            }
        });
    }
    
    public SetEnabled(enabled: boolean): void {
        this.enabled = enabled;
    }
    
    public Destroy(): void {
        if (!this.destroyed) {
            this.OnDestroy();
        }
    }
    
    public AddAwakeCallback(callback: (self: MonoBehaviour) => void): void {
        this.awakeCallbacks.push(callback);
    }
    
    public AddStartCallback(callback: (self: MonoBehaviour) => void): void {
        this.startCallbacks.push(callback);
    }
    
    public AddUpdateCallback(callback: (self: MonoBehaviour, deltaTime: number) => void): void {
        this.updateCallbacks.push(callback);
    }
    
    public AddFixedUpdateCallback(callback: (self: MonoBehaviour, fixedDeltaTime: number) => void): void {
        this.fixedUpdateCallbacks.push(callback);
    }
    
    public AddLateUpdateCallback(callback: (self: MonoBehaviour, deltaTime: number) => void): void {
        this.lateUpdateCallbacks.push(callback);
    }
    
    public AddDestroyCallback(callback: (self: MonoBehaviour) => void): void {
        this.destroyCallbacks.push(callback);
    }
    
    public SetGameObject(robloxObject: Instance): void {
        print(`[MonoBehaviour] SetGameObject - 接收到robloxObject: ${robloxObject ? robloxObject.Name : "nil"}`);
        if (!robloxObject) {
            warn("SetGameObject: robloxObject is nil, skipping GameObject creation");
            return;
        }
        this.gameObject = RobloxUnityAdapter.GameObject.create(robloxObject);
        print(`[MonoBehaviour] SetGameObject - 创建的gameObject: ${this.gameObject ? "成功" : "失败"}`);
        this.transform = this.gameObject!.transform;
        print(`[MonoBehaviour] SetGameObject - 创建的transform: ${this.transform ? "成功" : "失败"}`);
        print(`[MonoBehaviour] SetGameObject - 验证this.gameObject存在: ${this.gameObject !== undefined}`);
    }
    
    public GetComponent(componentType: string): any {
        if (this.gameObject) {
            return this.gameObject.GetComponent(componentType);
        }
        return undefined;
    }
    
    public GetComponentInChildren(componentType: string): Instance | undefined {
        if (this.gameObject) {
            return this.gameObject.GetComponentInChildren(componentType);
        }
        return undefined;
    }
    
    public GetComponentsInChildren(componentType: string): Instance[] {
        if (this.gameObject) {
            return this.gameObject.GetComponentsInChildren(componentType);
        }
        return [];
    }
    
    public AddComponent(componentType: string): Instance | undefined {
        if (this.gameObject) {
            return this.gameObject.AddComponent(componentType);
        }
        return undefined;
    }
}